filebeat/input/file/comparator.go

```diff
@@ -0,0 +1,211 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package file
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/file"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+const (
+	inodeDeviceIDName = "inode_deviceid"
+	pathName          = "path"
+	inodeMarkerName   = "inode_marker"
+
+	DefaultComparatorName = inodeDeviceIDName
+	comparatorNameSuffix  = "_comparator"
+)
+
+var (
+	comparatorFactories = map[string]identityComparatorFactory{
+		inodeDeviceIDName: newINodeDevideIDComparator,
+		pathName:          newPathComparator,
+		inodeMarkerName:   newINodeMarkerComparator,
+	}
+)
+
+// IdentityComparator identifies and compares files states.
+type IdentityComparator interface {
+	// IsSameFile determines if two states belong to the same file.
+	IsSameFile(*State, *State) bool
+	// IsEmptyState checks if state information is initialized.
+	IsEmptyState(*State) bool
+
+	// Name TODO
+	Name() string
+
+	// StateIdentifier returns something
+	StateIdentifier() StateIdentifier
+}
+
+type identityComparatorFactory func(*common.Config) (IdentityComparator, error)
+
+// NewIdentityComparator creates a new file identity comparator for an input.
+func NewIdentityComparator(c *common.Config) (IdentityComparator, error) {
+	cmpTypes := c.GetFields()
+	if len(cmpTypes) != 1 {
+		return nil, fmt.Errorf("invalid number of fields for identity_comparator, need one, got %v", cmpTypes)
+	}
+
+	cmpType := cmpTypes[0]
+	logp.Info(common.DebugString(c, false))
+	f, ok := comparatorFactories[cmpType]
+	if !ok {
+		return nil, fmt.Errorf("no such file_identity comparator: %s", cmpType)
+	}
+
+	c, err := c.Child(cmpType, -1)
+	if err != nil {
+		return nil, fmt.Errorf("error while transforming file_identity configuration")
+	}
```

Comment:
check out `common.ConfigNamespace` type. This type implements common logic for this style of configuration.